### PR TITLE
Move index meta information

### PR DESCRIPTION
**Pull request by muff1nman**
Move last scan data into index metadata. Reasoning:

1. Frequently changing data should not be in the properties file
2. Consistency can be improved by keeping the information about the index and the index data itself together
3. Should solve #1234 

This is still a WIP as it needs some cleaning up and testing.

### CODE REVIEW

**Comment by muff1nman** [comment]
@tesshucom would like to get any feedback you have on this. thanks.

**Comment by tesshucom** [comment]
It might have been confusing because I used the word atomic.
In fact, the processing done in doScanLibrary is somewhat robust with respect to atomicity.
So we can safely write statics within this block.
Legacy Server output statistics to a file because it was the most reliable method at that time.

But so both DB and Lucene have improved reliability at Airsonic, now we can save it both DB and Lucene .
Even if you scan from different JVMs, both DB and Lucene will eliminate double writing with its own lock.
(When saving to Lucene, simply change the method name to a more descriptive name.)

If we want to be safer, we should do the following:

https://github.com/tesshucom/airsonic/commit/cd8b5f4b358806dca415741cbb6cc08f8d2e6a81

___

It is off topic from here.
(But a little important)

The atomicity is uncertain when looking at the entire MediaScannerService.
In particular, isScanning () is ``syncronized``, but is vulnerable to multithreading.

When we call 10 times almost simultaneously, we expect the first  and others to return different values.
If we actually do it, it is not.

https://github.com/tesshucom/airsonic/commit/5127a33a8e796417dd0034b683fafc8210d29cdf

Any logic that relies on isScanning () is uncertain.
It is impossible to describe an exact monitoring process.
Therefore, it is correct in the direction to describe directly in doScanLibrary like this PR.

The original author probably knew this.
Since it is impossible to synchronize isScanning (), it looks like a policy that supports callback for all processing.
Since jsonp is supported, monitoring processing using JavaScript may be described in a better way than Airsonic.

**Comment by tesshucom** [comment]
It might have been confusing because I used the word atomic.
In fact, the processing done in doScanLibrary is somewhat robust with respect to atomicity.
So we can safely write statics within this block.
Legacy Server output statistics to a file because it was the most reliable method at that time.

But so both DB and Lucene have improved reliability at Airsonic, now we can save it both DB and Lucene .
Even if you scan from different JVMs, both DB and Lucene will eliminate double writing with its own lock.
(When saving to Lucene, simply change the method name to a more descriptive name.)

If we want to be safer, we should do the following:

https://github.com/tesshucom/airsonic/commit/cd8b5f4b358806dca415741cbb6cc08f8d2e6a81

___

It is off topic from here.
(But a little important)

The atomicity is uncertain when looking at the entire MediaScannerService.
In particular, isScanning () is ``syncronized``, but is vulnerable to multithreading.

When we call 10 times almost simultaneously, we expect the first  and others to return different values.
If we actually do it, it is not.

https://github.com/tesshucom/airsonic/commit/5127a33a8e796417dd0034b683fafc8210d29cdf

Any logic that relies on isScanning () is uncertain.
It is impossible to describe an exact monitoring process.
Therefore, it is correct in the direction to describe directly in doScanLibrary like this PR.

The original author probably knew this.
Since it is impossible to synchronize isScanning (), it looks like a policy that supports callback for all processing.
Since jsonp is supported, monitoring processing using JavaScript may be described in a better way than Airsonic if we do not consider safety.

**Comment by tesshucom** [comment]
It might have been confusing because I used the word atomic.
In fact, the processing done in doScanLibrary is somewhat robust with respect to atomicity.
So we can safely write statics within this block.
Legacy Server output statistics to a file because it was the most reliable method at that time.

But so both DB and Lucene have improved reliability at Airsonic, now we can save it both DB and Lucene .
Even if you scan from different JVMs, both DB and Lucene will eliminate double writing with its own lock.
(When saving to Lucene, simply change the method name to a more descriptive name.)

If we want to be safer, we should do the following:

https://github.com/tesshucom/airsonic/commit/cd8b5f4b358806dca415741cbb6cc08f8d2e6a81

___

It is off topic from here.
(But a little important)

The atomicity is uncertain when looking at the entire MediaScannerService.
In particular, isScanning () is ``syncronized``, but is vulnerable to multithreading.

When we call 10 times almost simultaneously, we expect the first  and others to return different values.
If we actually do it, it is not.

https://github.com/tesshucom/airsonic/commit/5127a33a8e796417dd0034b683fafc8210d29cdf

Any logic that relies on isScanning () is uncertain.
It is impossible to describe an exact monitoring process.
Therefore, it is correct in the direction to describe directly in doScanLibrary like this PR.

**Comment by tesshucom** [comment]
It might have been confusing because I used the word atomic.
In fact, the processing done in doScanLibrary is somewhat robust with respect to atomicity.
So we can safely write statics within this block.
Legacy Server output statistics to a file because it was the most reliable method at that time.

But so both DB and Lucene have improved reliability at Airsonic, now we can save it both DB and Lucene .
Even if you scan from different JVMs, both DB and Lucene will eliminate double writing with its own lock.
(When saving to Lucene, simply change the method name to a more descriptive name.)

If we want to be safer, we should do the following:
(However, if we write seriously, shutdown processing is also required)

https://github.com/tesshucom/airsonic/commit/cd8b5f4b358806dca415741cbb6cc08f8d2e6a81



___

It is off topic from here.
(But a little important)

The atomicity is uncertain when looking at the entire MediaScannerService.
In particular, isScanning () is ``syncronized``, but is vulnerable to multithreading.

When we call 10 times almost simultaneously, we expect the first  and others to return different values.
If we actually do it, it is not.

https://github.com/tesshucom/airsonic/commit/5127a33a8e796417dd0034b683fafc8210d29cdf

Any logic that relies on isScanning () is uncertain.
It is impossible to describe an exact monitoring process.
Therefore, it is correct in the direction to describe directly in doScanLibrary like this PR.

**Comment by muff1nman** [comment]
I think we can fix synchronization issues in another pr.

**Comment by muff1nman** [comment]
Not sure where travis ran off to, but the build is passing on my machine.

File: airsonic-main/src/main/java/org/airsonic/player/service/search/IndexManager.java [hunk 1]
```diff
@@ -203,6 +205,42 @@ private void stopIndexing(IndexType type) {
 
     }
 
+    /**
+     * Return the MediaLibraryStatistics saved on commit in the index. Ensures that each index reports the same data.
+     * On invalid indices, returns null.
+     */
+    public @Nullable MediaLibraryStatistics getStatistics() {
```
[the remark this answers is not in the document]
Comment (muff1nman) [reply airsonic-main/src/main/java/org/airsonic/player/service/search/IndexManager.java]: The database would be better than the current properties file, but doesn't have the nice behavior that the lastscanned automatically gets removed if you move your index files.

**Pull request merged**